File: LeetCode/Python/102.binary-tree-level-order-traversal.py
# @before-stub-for-debug-begin
from python3problem102 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=102 lang=python3
#
# [102] Binary Tree Level Order Traversal
#

# @lc code=start
# Definition for a binary tree node.
# """
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
# """
# TODO: Implement
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return
        
        queue = [root]
        lst = [[root]]
        
        temp_root = root
        run = 1
        while queue:
            temp_queue = []
            
            if queue[0].left:
                queue.append(queue[0].left)
                temp_queue.append(queue[0].left)
            
            if queue[0].right:
                queue.append(queue[0].right)
                temp_queue.append(queue[0].right)
            
            lst.append(temp_queue)
            queue.pop(0)
            # if temp_root.left:
            #     queue.append(temp_root.left)
            
            # if temp_root.right:
            #     queue.append(temp_root.right)
        
        # lst.append([root.left.val, root.right.val] + self.levelOrder(root.right) + self.levelOrder(root.right))
        # print(lst)
        return lst


# @lc code=end

File: LeetCode/CPP/102.binary-tree-level-order-traversal.cpp
// @before-stub-for-debug-begin
#include <vector>
#include <string>
#include "commoncppproblem102.h"

using namespace std;
// @before-stub-for-debug-end

/*
 * @lc app=leetcode id=102 lang=cpp
 *
 * [102] Binary Tree Level Order Traversal
 */
#include <bits/stdc++.h>
using namespace std;

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// struct TreeNode
// {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };

class Solution
{
public:
    vector<vector<int>> levelOrder(TreeNode *root)
    {
        vector<vector<int>> arr;
        queue<TreeNode *> q;
        if (root == nullptr)
        {
            return arr;
        }
        q.push(root);

        while (!q.empty())
        {
            vector<int> v;
            int level = q.size();
            for (int i = 0; i < level; i++)
            {
                v.push_back(q.front()->val);
                if (q.front()->left)
                {
                    q.push(q.front()->left);
                }
                if (q.front()->right)
                {
                    q.push(q.front()->right);
                }
                q.pop();
            }
            arr.push_back(v);
        }
        return arr;
    }
};
// @lc code=end